processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java

```diff
@@ -353,31 +383,51 @@ public String getOutOfRowsReason()
   @Override
   public float getMetricFloatValue(int rowOffset, int aggOffset)
   {
-    return concurrentGet(rowOffset)[aggOffset].getFloat();
+    if (isPreserveExistingMetrics()) {
+      return (float) getCombinedMetric(getMetricAggs(), concurrentGet(rowOffset), aggOffset);
+    } else {
+      return concurrentGet(rowOffset)[aggOffset].getFloat();
+    }
   }
 
   @Override
   public long getMetricLongValue(int rowOffset, int aggOffset)
   {
-    return concurrentGet(rowOffset)[aggOffset].getLong();
+    if (isPreserveExistingMetrics()) {
+      return (long) getCombinedMetric(getMetricAggs(), concurrentGet(rowOffset), aggOffset);
+    } else {
+      return concurrentGet(rowOffset)[aggOffset].getLong();
+    }
   }
 
   @Override
   public Object getMetricObjectValue(int rowOffset, int aggOffset)
   {
-    return concurrentGet(rowOffset)[aggOffset].get();
+    if (isPreserveExistingMetrics()) {
+      return getCombinedMetric(getMetricAggs(), concurrentGet(rowOffset), aggOffset);
+    } else {
+      return concurrentGet(rowOffset)[aggOffset].get();
+    }
   }
 
   @Override
   protected double getMetricDoubleValue(int rowOffset, int aggOffset)
   {
-    return concurrentGet(rowOffset)[aggOffset].getDouble();
+    if (isPreserveExistingMetrics()) {
+      return (double) getCombinedMetric(getMetricAggs(), concurrentGet(rowOffset), aggOffset);
+    } else {
+      return concurrentGet(rowOffset)[aggOffset].getDouble();
+    }
```

Comment:
This `if/else` logic repeats itself. It makes sense to put it in a function or move the entire logic to `getCombinedMetric()`.

Reply:
Done